astropy/wcs/src/wcslib_celprm_wrap.c

```diff
@@ -0,0 +1,486 @@
+
+
+static PyObject* PyCelprm_get_theta0(PyCelprm* self, void* closure)
+{
+    if (is_cel_null(self)) {
+        return NULL;
+    } else if (self->x->theta0 != UNDEFINED) {
+        return get_double("theta0", self->x->theta0);
+    }
+    Py_RETURN_NONE;
+}
+
+
+static int PyCelprm_set_theta0(PyCelprm* self, PyObject* value, void* closure)
+{
+    int result;
+    double theta0;
+    if(is_cel_null(self) || is_readonly(self)) {
+        return -1;
+    } else if (value == Py_None) {
+        if (self->x->theta0 != UNDEFINED) {
+            self->x->theta0 = UNDEFINED;
+            self->x->flag = 0;
+        }
+    } else {
+        result = set_double("theta0", value, &theta0);
+        if (result) return result;
+        if (theta0 != self->x->theta0) {
+            self->x->theta0 = theta0;
+            self->x->flag = 0;
+        }
+    }
+    return 0;
+}
+
+
+static PyObject* PyCelprm_get_ref(PyCelprm* self, void* closure)
+{
+    Py_ssize_t size = 4;
+    if (is_cel_null(self)) {
+        return NULL;
+    } else {
+        return get_double_array("ref", self->x->ref, 1, &size, (PyObject*) self);
+    }
+}
+
+
+static int PyCelprm_set_ref(PyCelprm* self, PyObject* value, void* closure)
+{
+    int i;
+    int skip[4] = {0, 0, 0, 0};
+    double ref[4] = {0.0, 0.0, UNDEFINED, +90.0};
+    npy_intp size;
+    double *data;
+
+    if (is_cel_null(self) || is_readonly(self)) return -1;
+
+    if (value == Py_None) {
+        /* If ref is set to None - reset ref to celini values: */
+        for (i = 0; i < 4; i++) {
+            self->x->ref[i] = ref[i];
+        }
+        self->x->flag = 0;
+        return 0;
+    }
+
+    PyObject* value_array = PyArray_ContiguousFromAny(value, NPY_DOUBLE, 1, 1);
+    if (!value_array) return -1;
+
+    size = PyArray_SIZE(value_array);
+
+    if (size < 1) {
+        Py_DECREF(value_array);
+        PyErr_SetString(PyExc_ValueError,
+            "'ref' must be a non-empty 1-dimentional list of values or None.");
+        return -1;
+    }
+
+    if (size > 4) {
+        Py_DECREF(value_array);
+        PyErr_SetString(PyExc_RuntimeError, "Number of 'ref' values cannot exceed 4.");
+        return -1;
+    }
+
+    if (PyList_Check(value)) {
+        for (i = 0; i < size; i++) {
+            skip[i] = (PyList_GetItem(value, i) == Py_None);
+        }
+    }
+
+    data = (double*) PyArray_DATA(value_array);
+
+    for (i = 0; i < size; i++) {
+        if (skip[i]) continue;
+        self->x->ref[i] = data[i];
+        if (npy_isnan(self->x->ref[i])) {
+            self->x->ref[i] = UNDEFINED;
+        } else {
+            self->x->ref[i] = data[i];
+        }
+    }
+    for (i = size; i < 4; i++) {
+        self->x->ref[i] = ref[i];
+    }
+
+    self->x->flag = 0;
+    Py_DECREF(value_array);
+    return 0;
+}
+
+
+static PyObject* PyCelprm_get_prj(PyCelprm* self, void* closure)
+{
+    if (is_cel_null(self)) return NULL;
+    return (PyObject*)PyPrjprm_cnew((PyObject *)self, &(self->x->prj), NULL);
```

Comment:
Does this return a new Python object every time the getter is accessed?  So that this fails:

```python
assert cel.prj == cel.prj
```

Is that desirable?

Reply:
I wouldn't say it is desirable but I also do not see this as a big issue. I do not see why would anyone try to check `cel.prj == cel.prj`. Since all other members of `Wcsprm` are accessed like this, I would suggest to leave this "as is" for now and fix it in a separate PR that would address all other "sub-structures" (`Celprm`, `Prjprm`, `Aux`, `Tabprm`, and `Wtbarr`).

Reply:
I wouldn't say it is desirable but I also do not see this as a big issue. I do not see why would anyone try to check `cel.prj is cel.prj`. With regard to `==` I could probably add `__eq__`. Since all other members of `Wcsprm` are accessed like this (one gets a new object), I would suggest to leave this "as is" for now and fix it in a separate PR that would address all other "sub-structures" (`Celprm`, `Prjprm`, `Aux`, `Tabprm`, and `Wtbarr`).

Reply:
However, the `is` does not work (i.e., produces surprising results) for other members of `Wcsprm` structure such as lists and arrays. For example,
```python
In [1]: from astropy import wcs

In [2]: w = wcs.WCS()

In [3]: w.wcs.ctype is w.wcs.ctype
Out[3]: False
```
So, for consistency, maybe it is better to leave things as is.

Reply:
Yes, definitely in favor of consistency!  I'll resolve this.

Reply:
But this is something that could be addressed in a separate PR if there are serious side-effects from this behavior affecting users. I doubt there will be since access to members is either to set something or to retrieve some information. `is` is mostly of theoretical interest, I think.